app/database/session_store.go

```diff
@@ -32,3 +32,47 @@ func (s *SessionStore) GetUserAndSessionIDByValidAccessToken(token string) (user
 
 	return result.User, result.SessionID, err
 }
+
+type sessionWithMostRecentIssuedAt struct {
+	SessionID int64
+	IssuedAt  Time
+}
+
+// GetUserSessionsSortedByMostRecentIssuedAt returns the user's sessions sorted by most recent issued_at.
+func (s *SessionStore) GetUserSessionsSortedByMostRecentIssuedAt(userID int64) []sessionWithMostRecentIssuedAt {
+	var sessions []sessionWithMostRecentIssuedAt
+
+	err := s.
+		Select(`
+			access_tokens.session_id AS session_id,
+			MAX(access_tokens.issued_at) AS issued_at
+		`).
+		Joins("JOIN access_tokens ON access_tokens.session_id = sessions.session_id").
+		Where("sessions.user_id = ?", userID).
+		Group("access_tokens.session_id").
+		Order("issued_at DESC").
+		Scan(&sessions).
+		Error()
+	mustNotBeError(err)
+
+	return sessions
+}
+
+// DeleteOldSessionsToKeepMaximum deletes old sessions to keep the maximum number of sessions.
+func (s *SessionStore) DeleteOldSessionsToKeepMaximum(userID int64, max int) {
+	sessions := s.GetUserSessionsSortedByMostRecentIssuedAt(userID)
+
+	if len(sessions) > max {
+		// Delete the oldest sessions.
+		oldestSessions := sessions[max:]
+
+		oldestSessionIDs := make([]int64, len(oldestSessions))
+		for i, session := range oldestSessions {
+			oldestSessionIDs[i] = session.SessionID
+		}
+
+		err := s.Delete("session_id IN (?)", oldestSessionIDs).
+			Error()
+		mustNotBeError(err)
```

Comment:
- Is there a reason to do a select (with `max` unused rows) and N delete queries ? If no, do the whole thing in one single SQL DELETE query!
- as discussed, make sure you delete sessions with no access token
- use a tie-breaker so that the rows you delete is predictable

Reply:
- The sessions without access tokens are now deleted in priority. Used `UNION ALL`, not sure if it's possible in a different way: on one end we have those with no join (left join), and on the other end we have is an aggregation.
- Added `session_id` as a tie-breaker

For the first item I'm not sure I understand. There's only one `DELETE` query with the `IN` operator. Do you mean merging the `SELECT` and the `DELETE` together? This might end with a query that's very difficult to understand.

Reply:
Discussion continued in Slack.

Reply:
fixed, the queries have been merged. There are some implementations details I've put in comments and in the commit messages (the transformation was done step-by-step in the last 3 commits).